packages/drpc/src/DrpcApi.ts

```diff
@@ -0,0 +1,152 @@
+import type { DrpcMessageRecord } from './repository/DrpcMessageRecord'
+import type { ConnectionRecord } from '@credo-ts/core'
+
+import {
+  AgentContext,
+  MessageHandlerRegistry,
+  MessageSender,
+  OutboundMessageContext,
+  injectable,
+  ConnectionService,
+} from '@credo-ts/core'
+
+import { DrpcRequestHandler, DrpcResponseHandler } from './handlers'
+import { DrpcService } from './services'
+
+@injectable()
+export class DrpcApi {
+  private drpcMessageService: DrpcService
+  private messageSender: MessageSender
+  private connectionService: ConnectionService
+  private agentContext: AgentContext
+
+  public constructor(
+    messageHandlerRegistry: MessageHandlerRegistry,
+    drpcMessageService: DrpcService,
+    messageSender: MessageSender,
+    connectionService: ConnectionService,
+    agentContext: AgentContext
+  ) {
+    this.drpcMessageService = drpcMessageService
+    this.messageSender = messageSender
+    this.connectionService = connectionService
+    this.agentContext = agentContext
+    this.registerMessageHandlers(messageHandlerRegistry)
+  }
+
+  /**
+   * sends the request object to the connection and returns a function that will resolve to the response
+   * @param connectionId the connection to send the request to
+   * @param request the request object
+   * @returns curried function that waits for the response
+   */
+  public async sendRequest(connectionId: string, request: DrpcRequest): Promise<() => Promise<DrpcResponse>> {
+    const connection = await this.connectionService.getById(this.agentContext, connectionId)
+    const { message: drpcMessage, record: drpcMessageRecord } = await this.drpcMessageService.createRequestMessage(
+      this.agentContext,
+      request,
+      connection.id
+    )
+    const messageId = drpcMessage.id
+    await this.sendMessage(connection, drpcMessage, drpcMessageRecord)
+    return this.recvResponse.bind(this, messageId)
+  }
+
+  /**
+   * Listen for a response that has a thread id matching the provided messageId
+   * @param messageId the id to match the response to
+   * @returns the response object
+   */
+  private async recvResponse(messageId: string): Promise<DrpcResponse> {
+    return new Promise((resolve) => {
+      const listener = ({
+        drpcMessageRecord,
+        removeListener,
+      }: {
+        drpcMessageRecord: DrpcMessageRecord
+        removeListener: () => void
+      }) => {
+        const message = drpcMessageRecord.message
+        if (drpcMessageRecord.threadId === messageId) {
+          removeListener()
+          resolve(message as DrpcResponse)
+        }
+      }
+
+      this.drpcMessageService.createResponseListener(listener)
+    })
+  }
+
+  /**
+   * Listen for a request and returns the request object and a function to send the response
+   * @returns the request object and a function to send the response
+   */
+  public async recvRequest(): Promise<{
```

Comment:
This naming reminds me socket's recv() call, so I guess the intention is to be used like that. In this case, to also mimic that function, I think it will be nice to make it possible to define an optional timeout so it's not forced to block